File: src/main/java/hdzi/editstarters/dependency/StarterInfo.java
package hdzi.editstarters.dependency;

import lombok.Getter;
import lombok.Setter;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;

@Getter
@Setter
public class StarterInfo extends Dependency {
    private String id;
    private String name;

    private String description;
    private String versionRange;

    private final List<Repository> repositories = new ArrayList<>();
    private Bom bom;

    public void addRepository(String id, Repository repository) {
        if (repository != null && !Points.contains(this.repositories, repository)) {
            repository.setId(id);
            this.repositories.add(repository);
        }
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        StarterInfo that = (StarterInfo) o;
        return Objects.equals(id, that.id);
    }

    @Override
    public int hashCode() {
        return Objects.hash(id);
    }

    @Override
    public String toString() {
        return this.name;
    }
}

File: src/main/java/hdzi/editstarters/bean/StarterInfo.kt
package hdzi.editstarters.bean

import hdzi.editstarters.bean.initializr.InitializrBom
import hdzi.editstarters.bean.initializr.InitializrRepository

/**
 * Created by taojinhou on 2018/12/21.
 */
class StarterInfo : Dependency {
    // 基本信息
    var id: String? = null
    var name: String? = null
    var description: String? = null
    var versionRange: String? = null
    // 坐标信息
    override var groupId: String? = null
    override var artifactId: String? = null
    var scope: String? = null
    var version: String? = null
    var repositories = mutableSetOf<InitializrRepository>()
    var bom: InitializrBom? = null
    var exist = false

    val searchKey: String
        get() = "$groupId\t$artifactId\t$name\t$description".toLowerCase()

    val canBeAdded: Boolean
        get() = groupId != null && artifactId != null

    fun addRepository(repository: InitializrRepository?) {
        if (repository != null) {
            this.repositories.add(repository)
        }
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as StarterInfo

        if (id != other.id) return false

        return true
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: 0
    }

    override fun toString(): String {
        return name!!
    }
}
